drivers/i3c/i3c_npcx.c

```diff
@@ -1835,57 +1924,359 @@ static int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc
+}
+
+static void npcx_i3c_target_enable_mdmatb(const struct device *dev, uint8_t *buf, uint16_t len)
+{
+	const struct npcx_i3c_config *config = dev->config;
+	struct i3c_reg *i3c_inst = config->base;
+	struct mdma_reg *mdma_inst = config->mdma_base;
+
+	/* Check MDMA disable */
+	if (IS_BIT_SET(mdma_inst->MDMA_CTL1, NPCX_MDMA_CTL_MDMAEN) != 0) {
+		mdma_inst->MDMA_CTL1 &= ~BIT(NPCX_MDMA_CTL_MDMAEN);
+		LOG_DBG("MDMATB_EN=1 before enable");
+	}
+
+	/* Detect a START and STOP only if the transaction address matches the target address */
+	i3c_inst->CONFIG |= BIT(NPCX_I3C_CONFIG_MATCHSS);
+
+	/* Enable DMA only for one frame.
+	 * MATCHSS must be set to 1 before selecting '0x1' for DMATB field
+	 *
+	 * In SDR DMATB is automatically cleared if MATCHED bit is set to 1 and either STOP bit
+	 * or START bit is set to 1.
+	 *
+	 * In HDR-DDR mode, DMATB is not automatically cleared.
+	 */
+	SET_FIELD(i3c_inst->DMACTRL, NPCX_I3C_DMACTRL_DMATB, MDMA_DMAFB_EN_ONE_FRAME);
+
+	/* Write Operation (MDMA CH_1) */
+	mdma_inst->MDMA_TCNT1 = len;                       /* Set MDMA transfer count */
+	mdma_inst->MDMA_SRCB1 = (uint32_t)buf;             /* Set source address */
+	mdma_inst->MDMA_CTL1 &= ~BIT(NPCX_MDMA_CTL_TC);    /* W0C */
+	mdma_inst->MDMA_CTL1 |= BIT(NPCX_MDMA_CTL_MDMAEN); /* Start DMA transfer */
+}
+
+static void npcx_i3c_target_rx_read(const struct device *dev)
+{
+	struct npcx_i3c_data *data = dev->data;
+	struct i3c_config_target *config_tgt = &data->config_target;
+
+	/* Enable the DMA from bus */
+	npcx_i3c_target_enable_mdmafb(dev, data->mdma_rx_buf, config_tgt->max_read_len);
+}
+
+/* brief:  Handle the end of transfer (read request or write request).
+ *         The ending signal might be either STOP or Sr.
+ * return: -EINVAL:
+ *              1. operation not read or write request.
+ *              2. start or stop flag is not set.
+ *         -EBUSY: in write request, wait for mdma done.
+ *          0: success
+ */
+static int npcx_i3c_target_xfer_end_handle(const struct device *dev)
+{
+	struct npcx_i3c_data *data = dev->data;
+	const struct npcx_i3c_config *config = dev->config;
+	struct i3c_reg *inst = config->base;
+	struct mdma_reg *mdma_inst = config->mdma_base;
+	const struct i3c_target_callbacks *target_cb = data->target_config->callbacks;
+	bool is_i3c_start = IS_BIT_SET(inst->INTMASKED, NPCX_I3C_INTMASKED_START);
+	bool is_i3c_stop = IS_BIT_SET(inst->INTMASKED, NPCX_I3C_INTMASKED_STOP);
+	enum npcx_i3c_oper_state op_state = get_oper_state(dev);
+	uint32_t cur_xfer_cnt;
+	uint32_t timer = 0;
+	int ret = 0;
+
+	if ((op_state != NPCX_I3C_OP_STATE_WR) && (op_state != NPCX_I3C_OP_STATE_RD)) {
+		LOG_ERR("%s: op_staste error :%d", __func__, op_state);
+		return -EINVAL;
+	}
+
+	if ((is_i3c_start | is_i3c_stop) == 0) {
+		LOG_ERR("%s: not the end of xfer, is_start: %d, is_stop:%d", __func__, is_i3c_start,
+			is_i3c_stop);
+		return -EINVAL;
+	}
+
+	if (get_oper_state(dev) == NPCX_I3C_OP_STATE_WR) { /* Write request */
```

Comment:
could you invert this condition, handle the other case with the existing code and a goto and unindent the rest?